File: ios/CoinFactory.swift
//
//  CoinFactory.swift
//  coinswallet
//
//  Created by Tal Grynbaum on 30/06/2023.
//

import Foundation
import WalletCore

class CoinFactory {
  private static var coinMap = [String: (String) -> Coin]()
  
  static func registerCoin(name: String, constructor: @escaping (String) -> Coin) {
    coinMap[name] = constructor
  }
  
  static func createCoin(coinName: String, mnemonic: String) -> Coin {
    print("in create coin", coinName, mnemonic)
    guard let constructor = coinMap[coinName.lowercased()] else {
      print("in create coin unsupported", coinName)
      fatalError("Unsupported coin: \(coinName)")
    }
    let coin =  constructor(mnemonic)
    print("after constructor coin: ", coin)
    return coin
  }
  class Coin {
    var wallet: HDWallet
    
    init(mnemonic: String) {
      print("mnemonic",mnemonic)
      guard let wallet = HDWallet(mnemonic: mnemonic, passphrase: "") else {
        fatalError("Failed to create HDWallet")
      }
      self.wallet = wallet
    }
    
    func getNewAddress() -> String {
      fatalError("Subclasses must override getNewAddress()")
    }
    
    func getPrivateKey() -> String {
      fatalError("Subclasses must override getPrivateKey()")
    }
    
    func signTransaction(rawData: String) -> String {
      fatalError("Subclasses must override signTransaction(rawData:)")
    }
  }
}

File: android/app/src/main/java/com/coinswallet/CoinFactory.java
// CoinFactory.java
package com.coinswallet;

import wallet.core.jni.*;
import java.util.*;
import java.util.function.Function;
import java.util.function.Supplier;
import android.util.Log;

public class CoinFactory {

    // private static final Map<String, Supplier<Coin>> coinMap = new HashMap<>();
    private static final Map<String, Function<String, Coin>> coinMap = new HashMap<>();

    public static void registerCoin(String name, Function<String, Coin> constructor) {
        Log.d("coinswallet", "before register: map size: " + coinMap.size());
        coinMap.put(name, constructor);
        Log.d("coinswallet", "after register: map size: " + coinMap.size());
    }

    public static Coin createCoin(String coinName, String mnemonic) {
        Log.d("coinswallet", "Creating coin: " + coinName);
        Log.d("coinswallet", "map size: " + coinMap.size());
        Function<String, Coin> constructor = coinMap.get(coinName.toLowerCase());
        Log.d("coinswallet", "factory.createCoin: ");
        Log.d("coinswallet", "factory.createCoin: " + constructor);
        if (constructor == null) {
            throw new IllegalArgumentException("Unsupported coin: " + coinName);
        }
        return constructor.apply(mnemonic);
    }

    public static abstract class Coin {
        protected HDWallet wallet;

        protected Coin(String mnemonic) {
            this.wallet = new HDWallet(mnemonic, "");
        }

        public abstract String getNewAddress();

        public abstract String getPrivateKey();

        public abstract String signTransaction(String rawData);
    }
}
